vertx-web/src/main/java/io/vertx/ext/web/templ/TemplateEngine.java

```diff
@@ -33,12 +33,17 @@
 public interface TemplateEngine {
 
   /**
-   * Render
+   * Render the template
+   * <p>
+   * <b>NOTE</b> if you call method directly (i.e. not using {@link io.vertx.ext.web.handler.TemplateHandler}) make sure
+   * that <i>templateFileName</i> is sanitized via {@link io.vertx.ext.web.impl.Utils#normalizePath(String)}
+   *
    * @param context  the routing context
-   * @param templateFileName  the template file name to use
+   * @param templateDirectory  the template directory to use
+   * @param templateFileName  the relative template file name to use
    * @param handler  the handler that will be called with a result containing the buffer or a failure.
    */
-  void render(RoutingContext context, String templateFileName, Handler<AsyncResult<Buffer>> handler);
+  void render(RoutingContext context, String templateDirectory, String templateFileName, Handler<AsyncResult<Buffer>> handler);
```

Comment:
Well its not only for inheritance; per example this makes it much easier to do A/B testing (different layouts at different paths).
Problem with keeping old method and passing _null_ for _templateDirectory_ to new one is that we would again need to somehow extract path to pass it to wrapped _loader_ object.
Maybe the better approach would be to keep #580 but deprecate "old" method, then remove it in some later version (3.5?). Yes, this gives some duplicate code but its much cleaner ... _TemplateHandlerImpl_ can already use the new call.